src/Storages/MergeTree/MergeTreeDataPartWriterWide.cpp

```diff
@@ -101,35 +103,95 @@ void MergeTreeDataPartWriterWide::write(const Block & block,
         fillIndexGranularity(index_granularity_for_block, block.rows());
     }
 
-    auto offset_columns = written_offset_columns ? *written_offset_columns : WrittenOffsetColumns{};
+    std::unique_ptr<ThreadPool> writing_thread_pool;
+    std::vector<WrittenOffsetColumns> offset_columns_per_column;
+    std::vector<ColumnWriteResult> write_results(columns_list.size());
+    if (settings.max_threads != 1)
+    {
+        offset_columns_per_column.reserve(columns_list.size());
+        writing_thread_pool = std::make_unique<ThreadPool>(settings.max_threads);
```

Comment:
Also using thread pool to write columns in parallel to regular filesystem looks overkill if the block is small enough (I guess it is OK if MergeTree uses S3 as the underlying storage)